### PR TITLE
Handle generics in methodimpls for default interface methods

### LINKED ISSUE #16354: [Default interfaces] VerifyVirtualMethodsImplemented wrong for generic interfaces
Type `Fooer` declares it implements ``IFoo`1<class [System.Runtime]System.ValueType>``, but it doesn't. We still load the type just fine, violating type safety in the process, because the logic that checks for default interface method implementation is bogus and succeeds if we implement any instantiation of that generic interface.

```
.assembly extern System.Runtime { }

.assembly notimplemented { }

.class interface private abstract auto ansi IFoo`1<T>
{
  .method public hidebysig newslot abstract virtual instance void Frob() cil managed
  {
  }
}

.class interface private abstract auto ansi IBar
       implements class IFoo`1<class [System.Runtime]System.String>
{
  .method public hidebysig newslot virtual final instance void Frob() cil managed
  {
    .override class IFoo`1<class [System.Runtime]System.String>::Frob
    ret
  }
}

.class private auto ansi beforefieldinit Fooer
       extends [System.Runtime]System.Object
       implements IBar, class IFoo`1<class [System.Runtime]System.ValueType>
{
  .method public hidebysig specialname rtspecialname 
          instance void .ctor() cil managed
  {
    ldarg.0
    call instance void [System.Runtime]System.Object::.ctor()
    ret
  }
}

.method public hidebysig static void Main() cil managed
{
  .entrypoint
  newobj instance void Fooer::.ctor()
  pop
  ret
}
```

### LINKED ISSUE #16355: [Default interfaces] FindDefaultInterfaceImplementation instantiates interfaces over bogus types
The logic in `MethodTable::FindDefaultInterfaceImplementation` around this code:

https://github.com/dotnet/coreclr/blob/24680bd7883a1f8424766c5196b96b803dfb796d/src/vm/methodtable.cpp#L7171-L7184

Uses the instantiation of the `IBar` interface in `pCurMT` (that is not generic) to load ``IFoo`1``. This should likely use a different strategy to iterate the methodimpls.

```
.assembly extern System.Runtime { }

.assembly unambiguousconstraint { }

.class interface private abstract auto ansi IFoo`1<T>
{
  .method public hidebysig newslot abstract virtual instance int32 Frob() cil managed
  {
  }
}

.class interface private abstract auto ansi IBar
       implements class IFoo`1<class [System.Runtime]System.Object>
{
  .method public hidebysig newslot virtual final instance int32 Frob() cil managed
  {
    .override class IFoo`1<class [System.Runtime]System.Object>::Frob
    ldc.i4.m1
    ret
  }
}

.class interface private abstract auto ansi IBaz
       implements class IFoo`1<class [System.Runtime]System.String>
{
  .method public hidebysig newslot virtual final instance int32 Frob() cil managed
  {
    .override class IFoo`1<class [System.Runtime]System.String>::Frob
    ldc.i4 100
    ret
  }
}

.class private auto ansi beforefieldinit Fooer
       extends [System.Runtime]System.ValueType
       implements IBar, IBaz
{
}

.class private auto ansi beforefieldinit Gen`1<(class IFoo`1<class [System.Runtime]System.String>) T>
       extends [System.Runtime]System.Object
{
  .method public hidebysig static int32 Check(!T x) cil managed
  {
    ldarga.s 0
    constrained. !T
    callvirt instance int32 class IFoo`1<class [System.Runtime]System.String>::Frob()
    ret
  }
}

.method public hidebysig static int32 Main() cil managed
{
  .entrypoint
  .locals init (valuetype Fooer)
  ldloc.0
  call int32 class Gen`1<valuetype Fooer>::Check(!0)
  ret
}
```

### LINKED ISSUE #20281: Wrong check for default interface implementation at type load time
Attempting to run crossgen on this results in a TypeLoadException because the runtime thinks we don't implement `IFoo<object>.Frob` when in fact we do. Next we'll likely hit #16355, but that's tracked separately.

```
.assembly extern System.Runtime { }

.assembly crossgendefault { }

.class interface private abstract auto ansi IFoo`1<T>
{
  .method public hidebysig newslot abstract virtual instance int32 Frob() cil managed
  {
  }
}

.class interface private abstract auto ansi IBar
       implements class IFoo`1<class [System.Runtime]System.Object>
{
  .method public hidebysig newslot virtual final instance int32 Frob() cil managed
  {
    .override class IFoo`1<class [System.Runtime]System.Object>::Frob
    ldc.i4 100
    ret
  }
}

.class private auto ansi beforefieldinit Fooer
       extends [System.Runtime]System.Object
       implements IBar
{
  .method public hidebysig specialname rtspecialname 
          instance void .ctor() cil managed
  {
    ldarg.0
    call instance void [System.Runtime]System.Object::.ctor()
    ret
  }
}

.method public hidebysig static int32 Main() cil managed
{
  .entrypoint
  newobj instance void Fooer::.ctor()
  callvirt instance int32 class IFoo`1<class [System.Runtime]System.Object>::Frob()
  ret
}
```

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
The existing logic looking at MethodImpls to find the default interface implementation was not correct in the presence of generics. The MethodImpl records we build during type load only contain inexact MethodDescs for the declMethod that are not suitable for comparisons.

The fix is to pipe through the token and resolve the declaring method from it at the time of dispatch.

Fixes #16355.
Fixes #20281.
Fixes #16354.

### CODE REVIEW

File: src/vm/methodtable.cpp [hunk 1]
```diff
@@ -7171,43 +7174,62 @@ BOOL MethodTable::FindDefaultInterfaceImplementation(
...
+                                    IfFailThrow(pMD->GetModule()->GetMDImport()->GetParentToken(it.GetToken(), &tkParent));
+
+                                    MethodTable* pDeclMT = ClassLoader::LoadTypeDefOrRefOrSpecThrowing(
+                                        pMD->GetModule(),
+                                        tkParent,
+                                        &typeContext).AsMethodTable();
+
+                                    // We do CanCastToInterface to also cover variance.
+                                    // We already know this is a method on the same type definition as the (generic)
+                                    // interface but we need to make sure the instantiations match.
```
Comment (davidwrighton) [inline src/vm/methodtable.cpp]: Variant dispatch is typically done via a two pass technique. Wherein we first iterate to see if there is an exact match, and then a second iteration through the interfaces (of a particular type) to attempt to do a variant dispatch. Looking at the logic in this function, it appears that the variant matching rules here may not be followed, and the variance test cases added do not appear to cover the exact/inexact matching issue.

Comment (MichalStrehovsky) [reply src/vm/methodtable.cpp]: The issue of it ignoring variant interface dispatch rules and not doing two passes is orthogonal to the thing I'm fixing here.

I would prefer to track that in a separate issue/pull request for accounting purposes.

Comment (MichalStrehovsky) [reply src/vm/methodtable.cpp]: #20452

**Comment by MichalStrehovsky** [comment]
@dotnet-bot test this please

**Comment by MichalStrehovsky** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Pull request merged**